pkg/operator/starter.go

```diff
@@ -479,10 +484,16 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		controllerContext.EventRecorder,
 		os.Getenv("IMAGE"),
 		os.Getenv("OPERATOR_IMAGE"),
-		featureGateAccessor)
+		featureGateAccessor,
+		etcdBackupInformer,
+		jobsInformer)
 
+	go jobsInformer.Informer().Run(ctx.Done())
+	go etcdBackupInformer.Informer().Run(ctx.Done())
+	go configBackupInformer.Informer().Run(ctx.Done())
```

Comment:
Perhaps we can conditionally run these with the feature gate so the informer doesn't error out if the CRDs don't exist?
See e.g:
https://github.com/openshift/insights-operator/blob/0338819de275e2854d68df955c3d7beceb7f5fa7/pkg/controller/operator.go#L131C5-L141

Reply:
@hasbro17 hacked this by just faking the HasSynced functions. Works well :package:  

> Perhaps we can conditionally run these with the feature gate so the informer doesn't error out if the CRDs don't exist?

that's also an alternative way, but then we would need to restart the entire operator when the feature flags are enabled/disabled. Otherwise it becomes a bit messy to manage the informer routines. 

Happy to also code that approach for further discussion.

Reply:
Haha that's pretty clever 😄 

My only slight concern in bypassing/disabling `HasSynced()` is whether there would be any unintended side effects. Does this handle runtime enablement of the feature flag (i.e CRD appearing) gracefully? As in, the cache is updated with the CR as a watch event. Or do we wait until it periodically resyncs every 1 min? If not, then this is fine.

> that's also an alternative way, but then we would need to restart the entire operator when the feature flags are enabled/disabled. Otherwise it becomes a bit messy to manage the informer routines.

Oh yeah I definitely don't want us handling or restarting the informer routines at runtime. I think restarting the operator after enabling the feature flags is probably not too much to ask so we can do a clean sync of the new APIs.
I actually expected that would happen automatically based on:
> // By default, when the enabled/disabled list  of featuregates changes, os.Exit is called.

which should cause the operator's main to terminate as well (or so I think from a [quick test](https://go.dev/play/p/xlsdNVzdzG_P)).

https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/featuregates/simple_featuregate_reader.go#L87
https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/featuregates/simple_featuregate_reader.go#L151-L155

Does our operator not restart automatically if the feature gates change?

Reply:
We can keep the `HasSynced ` method if it turns out the operator isn't automatically restarting on feature gate changes. We can address that issue later. Just surprised at what other folks might be doing for their feature gates.

Reply:
>  Does this handle runtime enablement of the feature flag (i.e CRD appearing) gracefully? As in, the cache is updated with the CR as a watch event. 

Yes it does, the informer routine basically runs in the background the whole time and will react normal once the API is available. The drawback is that it has error messages in the logs once in a while. 

> Or do we wait until it periodically resyncs every 1 min? 

Both. The controller will still re-sync based on the schedule and once the informer is finding the API it will trigger the sync as usual, too.

>  I think restarting the operator after enabling the feature flags is probably not too much to ask so we can do a clean sync of the new APIs.

That's fair, that will also fix it. Currently the exit handler is overruled by one that just prints the change:
https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/starter.go#L166-L169

> Does our operator not restart automatically if the feature gates change?

Nope. The controller only runs when the feature flag is found to be enabled, so we can get away entirely without restarts. Maybe the restart solution would be much cleaner and we could gate both informers and controllers directly with the gate.

Reply:
here's the alternative with restarts:
https://github.com/openshift/cluster-etcd-operator/pull/1082

looks much cleaner to use, WDYT?

Reply:
Thanks for adding the alternate version! I think that looks much easier to understand and reason about. We can close this one in favor of that.